### PR TITLE
[CM-371] Various test fixes

### PR DESCRIPTION
**Pull request by mschuwalow**
https://liveintent.atlassian.net/browse/CM-371
Short description if any.

Author Todo List:

- [x] Add/adjust tests (if applicable)
- [x] Build in CI passes
- [x] Latest master revision is merged into the branch
- [x] Self-Review
- [x] Set `Ready For Review` status

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -14,7 +14,7 @@
     "node": ">=8"
   },
   "scripts": {
-    "test:unit": "npm run lint && nyc mocha --config test-config/mocha.unit.json --require @babel/register ",
+    "test:unit": "npm run lint && nyc mocha --config test-config/mocha.unit.json --require @babel/register --require @babel/polyfill",
```
Comment (mschuwalow) [inline package.json:17]: required for async

File: package.json [hunk 2]
```diff
@@ -38,13 +38,13 @@
...
     "@rollup/plugin-replace": "^2.3.3",
     "@rollup/plugin-strip": "^1.3.3",
-    "@wdio/browserstack-service": "5.13.2",
-    "@wdio/cli": "5.15.0",
-    "@wdio/junit-reporter": "5.15.0",
-    "@wdio/local-runner": "5.15.0",
-    "@wdio/mocha-framework": "5.15.0",
-    "@wdio/spec-reporter": "5.14.5",
-    "@wdio/sync": "5.14.4",
+    "@wdio/browserstack-service": "^6.12.0",
```
Comment (mschuwalow) [inline package.json:41]: version < 6.1.10 will not report status correctly to browserstack https://github.com/webdriverio/webdriverio/commit/85ee21d7aedb9ea1bcac0b754f5b19a799cbe1dc

File: package.json [hunk 3]
```diff
@@ -85,6 +85,6 @@
     "sinon": "^9.0.2",
     "sinon-chai": "^3.5.0",
     "tiny-uuid4": "^1.0.1",
-    "wdio-docker-service": "2.1.2"
+    "wdio-docker-service": "3.1.3"
```
Comment (mschuwalow) [inline package.json:88]: highest version that works with wdio v6 and doesn't require node 12

File: test-config/wdio.browserstack.conf.js [hunk 4]
```diff
@@ -6,7 +6,11 @@ const commonCapabilities = {
   'browserstack.console': 'verbose',
   'browserstack.debug': true,
   'browserstack.networkLogs': true,
-  'browserstack.appium_version': '1.14.0'
+  'browserstack.appium_version': '1.14.0',
+  timeouts: {
+    script: 60000,
```
Comment (mschuwalow) [inline test-config/wdio.browserstack.conf.js:11]: Internet explorer seems to default to super low timeouts

File: test-config/wdio.browserstack.conf.js [hunk 5]
```diff
@@ -167,8 +182,7 @@ exports.config = {
   // See the full list at http://mochajs.org/
   mochaOpts: {
     ui: 'bdd',
-    timeout: 90000,
-    require: ['@babel/register']
+    timeout: 600000,
+    require: ['@babel/register', '@babel/polyfill']
```
Comment (mschuwalow) [inline test-config/wdio.browserstack.conf.js:186]: required for async

File: test-config/wdio.docker.chrome.conf.js [hunk 6]
```diff
@@ -7,10 +7,11 @@ exports.config = {
   // WebdriverIO allows it to run your tests in arbitrary locations (e.g. locally or
   // on a remote machine).
   runner: 'local',
-  //
-  // Override default path ('/wd/hub') for chromedriver service.
-  // path: '/',
-  //
+
+  hostname: 'localhost',
```
Comment (mschuwalow) [inline test-config/wdio.docker.chrome.conf.js:11]: now required in version 6

File: test-config/wdio.docker.chrome.conf.js [hunk 7]
```diff
@@ -141,6 +137,6 @@ exports.config = {
   mochaOpts: {
     ui: 'bdd',
     timeout: 60000,
-    require: ['@babel/register']
+    require: ['@babel/register', '@babel/polyfill']
```
Comment (mschuwalow) [inline test-config/wdio.docker.chrome.conf.js:140]: for async

File: test-config/wdio.docker.chrome.conf.js [hunk 8]
```diff
@@ -96,11 +97,6 @@ exports.config = {
   //
   // Default request retries count
   connectionRetryCount: 3,
-  //
-  // Test runner services
-  // Services take over a specific job you don't want to take care of. They enhance
-  // your test setup with almost no effort. Unlike plugins, they don't add new
-  // commands. Instead, they hook themselves up into the test process.
   services: ['docker'],
   dockerOptions: {
```
Comment (mschuwalow) [inline test-config/wdio.docker.chrome.conf.js:101]: as this is a third party plugin the syntax didn't change in v6

File: test/it/helpers/browser.js [hunk 9]
```diff
@@ -1,102 +1,191 @@
...
+const WAIT_UNTIL_TIMEOUT_MILLIS = 30000
+const WAIT_UNTIL_INTERVAL = 500
+
+export async function getText (selector) {
+  var text
+  await browser.waitUntil(
+    async () => {
+      try {
+        const element = await $(selector)
+        if (element.elementId) {
```
Comment (mschuwalow) [inline test/it/helpers/browser.js:12]: Seems that after dom manipulation it might return an element with an undefined elementid. Trying to call getText on that will blow up

File: test/it/helpers/browser.js [hunk 10]
```diff
@@ -1,102 +1,191 @@
...
-      return server.getIdexHistory().length === expectedRequests
-    }, WAIT_UNTIL_TIMEOUT_MILLIS, 'resolveIdentity timed out', WAIT_UNTIL_INTERVAL)
-  } catch (e) {
-    console.error(e, server.getHistory().length, expectedRequests)
-  }
+export async function waitForRequests (expectedRequests, server) {
+  console.info(`Waiting for ${expectedRequests} requests`)
+  await browser.waitUntil(
+    () => {
+      return server.getHistory().length >= expectedRequests
```
Comment (mschuwalow) [inline test/it/helpers/browser.js:55]: tbh we could change this back to be an exact comparison

File: test/it/helpers/browser.js [hunk 11]
```diff
@@ -1,102 +1,191 @@
...
+  await browser.waitUntil(
+    async () => {
       try {
-        var key = 'x'
-        window.localStorage.removeItem(key)
-        window.localStorage.setItem(key, key)
-        return window.localStorage.getItem(key) === key
-      } catch (e) {
+        const idex = await $('#idex')
+        if (idex.elementId) {
```
Comment (mschuwalow) [inline test/it/helpers/browser.js:118]: the elementid being undefined fun again

File: test/it/live-connect.spec.js [hunk 12]
```diff
@@ -34,55 +34,58 @@ describe('LiveConnect', function () {
         ajaxTimeout: 3000
       }
     })
+    // wait for environment to become ready
+    await server.openPage('bln.test.liveintent.com', 'empty')
```
Comment (mschuwalow) [inline test/it/live-connect.spec.js:38]: turns out `window.localstorage` will be undefined until the first page load.
As we are probing localstorage before the first page load not doing this will cause the first test to always fail

File: test-config/wdio.browserstack.conf.js [hunk 13]
```diff
@@ -32,7 +36,6 @@ const allCapabilities = [
 ]
 
 exports.config = {
-  sync: true,
```
Comment (mschuwalow) [inline test-config/wdio.browserstack.conf.js:35]: doesn't seem to work correctly in all cases and is deprecated in version 7 anyway

File: test/it/helpers/browser.js [hunk 14]
```diff
@@ -122,3 +211,41 @@ export function isFirefoxAfter86 () {
   return browser.capabilities.browserName === 'firefox' &&
     parseInt(browser.capabilities.browserVersion.substring(0, 2)) > 86
 }
+
+export function redefineSetTimeout () {
```
Comment (mschuwalow) [inline test/it/helpers/browser.js:215]: Read the linked issue for details.

tldr: ios on browserstack signals that it's w3c spec compliant, but the wdio team doesn't give a shit. sooooo ... 👇

**Review by vmwhereqa — APPROVED** [review]

File: .circleci/config.yml [hunk 15]
```diff
@@ -30,13 +30,13 @@ references:
...
     run:
-      name: Install node v10.16.3
+      name: Install node v12
       command: |
         set +e
         export NVM_DIR="/opt/circleci/.nvm"
         [ -s "$NVM_DIR/nvm.sh" ] && \. "$NVM_DIR/nvm.sh"
-        nvm install v10.16.3
-        nvm alias default v10.16.3
+        nvm install v12.22.11
```
Comment (mschuwalow) [inline .circleci/config.yml:38]: required for v7

File: test-config/wdio.browserstack.conf.js [hunk 16]
```diff
@@ -1,38 +1,38 @@
...
-  { ...commonCapabilities, browserName: 'Safari', browser_version: '15.0', os_version: '15 Beta', device: 'iPhone 11 Pro', real_mobile: true },
-  { ...commonCapabilities, browserName: 'Safari', browser_version: '14.0', os_version: '14', device: 'iPhone 12', real_mobile: true },
-  { ...commonCapabilities, browserName: 'Safari', browser_version: '13.0', os_version: '13', device: 'iPhone 11', real_mobile: true },
-  { ...commonCapabilities, browserName: 'Safari', browser_version: '12.0', os_version: '12', device: 'iPhone 8', real_mobile: true },
-  { ...commonCapabilities, browserName: 'Safari', browser_version: '11.0', os_version: '11', device: 'iPad Pro 9.7 2016', real_mobile: true },
-  { ...commonCapabilities, browserName: 'Chrome', browser_version: 'latest', os_version: '11.0', device: 'Samsung Galaxy S21', real_mobile: true },
-  { ...commonCapabilities, browserName: 'Chrome', browser_version: '67', os_version: '6.0', device: 'Google Nexus 6', real_mobile: true }
+  { browserName: 'Chrome', browserVersion: 'latest', 'bstack:options': { ...commonBStackCapabilities, os: 'Windows', osVersion: '10'}},
```
Comment (mschuwalow) [inline test-config/wdio.browserstack.conf.js:12]: new format for v7

File: test/it/helpers/browser.js [hunk 17]
```diff
@@ -1,113 +1,206 @@
...
+  return browser.capabilities.browserName === 'safari' &&
+    (browser.capabilities.realMobile || browser.capabilities.real_mobile)
 }
 
 export function isMobileSafari14OrNewer () {
   return browser.capabilities.browserName === 'safari' &&
-    parseInt(browser.capabilities.version.substring(0, 2)) >= 14 &&
-    browser.capabilities.real_mobile
+    (
+      (browser.capabilities.browserVersion && parseInt(browser.capabilities.browserVersion.substring(0, 2)) >= 14) ||
```
Comment (mschuwalow) [inline test/it/helpers/browser.js:200]: These are a bit annoying now as they can be reported either w3c or in json web format.
Sadly the wdio heuristic (exposed via `browser.isW3C`) for deciding whether w3c is used is unreliable, so we have to check for both

**Review by 3link — APPROVED** [review]
